Elasticsearch json logging

pgomulka opened the pull request:
In order to log additional node.id and cluster.uuid a cluster state update will have to be received and the values passed to log4j pattern converter.
The implementation in this PR uses `ClusterStateObserver.Listener` to receive only one `ClusteStateUpdate`. Once update is received the  nodeId and clusterUUid are set in `NodeAndClusterIdConverter`. There was no need to use `ThreadLocal` as the pefromance gain compared to just `AtomicReference` (used in SetOnce) was not noticable. 

ESJsonLayout class is created to wrap the pattern we use to format json logs. Users are free to use their own patterns, but if they want smooth Beats integration and json format, they should use ESJsonLayout. `EvilLoggerTests` are left intact to make sure user's custom log patterns work fine.

As per initial issue we expect following fields in json log lines: type, tiemstamp, level, component, cluster.name, node.name, node.id, cluster.uuid, message, stacktrace
see `ESJsonLayout.java` for more details
- `type` field is defined in `log4j2.properties` as a layout parameter and has these values at the moment: `main` - for the rolling file, but renamed this as it won't make sens in docker, `deprecation`, `index_search_slowlog`, `index_indexing_slowlog`
- `stacktrace` field is a json string array. Each element in an array represent a line of a stactrace. If the Exception message or cause exception message is a multiline string (or json), then each line will be a separate message. Each stacktrace array element will be **as a new line** in logs.
- `message` field is a json escaped string - that means if the message is multiline it will end up in a single line with new line escaped.
- `timestamp` field is an ISO8601 with additional timezone ID

See sample log files in beats repository: `https://github.com/elastic/beats/pull/9988`

Docker log4j2 configuration is now almost the same as the one use for ES binary.  The only difference is that docker is using console appenders, whereas ES is using file appenders. 

The previous pattern appenders will be kept and will use `*.log` files. New appenders will use `.*json` files. Docker will only log in JSON

closes: #32850

pgomulka committed:
NodeId available on all log lines

Two aproaches. First with system variable and a plugin with atomic
reference. Second with custom Marker, LoggerContext, Selector and
Factory.

elasticmachine commented:
Pinging @elastic/es-core-infra

pgomulka commented on server/build.gradle:
```diff
@@ -123,6 +123,10 @@ dependencies {
   // repackaged jna with native bits linked against all elastic supported platforms
   compile "org.elasticsearch:jna:${versions.jna}"
 
+  compile "com.fasterxml.jackson.core:jackson-databind:${versions.jackson}"
```
I have abandoned the idea of using  Json layout for the moment. Not needed for current hardcoded pattern layout

nik9000 replied:
I'm quite weary of having this jar with the server's permissions. I *think* it is the one that usually sees security issues when Jackson gets security problems.

pgomulka replied:
Yes - I haven't decided to use it Json layout after all, will remove related code from the review. It was good to confirm that it is not what we want to follow, thanks for looking into this bit

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/MarkerLogger.java:
```diff
@@ -0,0 +1,181 @@
+import java.util.WeakHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class MarkerLogger extends Logger {
```
This is mostly a copy from PrefixLogger as it was ment to show the approach with using marker name as a placeholder for cluster_id and node_id

pgomulka commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -437,6 +441,14 @@ protected Node(
                                                  namedWriteableRegistry).stream())
                 .collect(Collectors.toList());
 
+            NodeIdPatternConverter nodeIdPatternConverter = NodeIdPatternConverter.newInstance(new String[]{});
```
very often I see ClusteStateListeners adding themselves to ClusterService with clusterService.addListener(this) inside a constructor. I don't think this a good idea as leaking the this reference during construction can cause race conditions and visibility problems.
I guess we should add them once constructed like here in 445

pgomulka commented on server/src/main/resources/org/elasticsearch/bootstrap/security.policy:
```diff
@@ -55,6 +55,11 @@ grant codeBase "${codebase.plugin-classloader}" {
+   permission java.util.PropertyPermission "*", "write";
+
+
+  permission java.lang.RuntimePermission "accessDeclaredMembers";
```
that was for Json layout. If we were to decide using it, we would have to leave a permission like this

nik9000 replied:
I'm super sure we don't want to grant this to the server in general. We've worked hard to remove it where possible.

pgomulka replied:
it will be gone - as discussed in [previous comment](https://github.com/elastic/elasticsearch/pull/36833#discussion_r242889744)

nik9000 commented on server/src/main/resources/org/elasticsearch/bootstrap/security.policy:
```diff
@@ -55,6 +55,11 @@ grant codeBase "${codebase.plugin-classloader}" {
 //// Everything else:
 
 grant {
+   permission java.util.PropertyPermission "*", "write";
```
I'm fairly sure we don't want to grant ourselves permission to do this.

pgomulka replied:
Yes that was just a shortcut. But will need to allow write property permission for node_id & cluster_id if we decide to use System.setProperty

pgomulka committed:
revert back changes due to use of json layout

pgomulka committed:
Merge branch 'master' into feature/logging-structured

pgomulka committed:
remove unused improts

pgomulka committed:
cleanup in plugin

pgomulka commented:
There are two other options that I will try investigate:
one would be to Modify and Reload log4j configuration with a new static pattern containing the node_id and clusterId once the ClusteStateChange is received.

another would be to write a custom formatter that will be checking MDC for presence of the node_id&cluster_id. If it was not yet set, it would try to get it from a shared AtomicReference and set in the MDC for future uses.

Both should options should be doable, but are not well documented in log4j docs.

pgomulka commented on distribution/src/config/jvm.options:
```diff
@@ -120,3 +120,5 @@ ${error.file}
 # temporary workaround for C2 bug with JDK 10 on hardware with AVX-512
 10-:-XX:UseAVX=2
+
+-Dlog4j2.loggerContextFactory=org.elasticsearch.common.logging.MarkerLoggerContextFactory
```
That will be very problematic in integration tests. Another argument against option3

pgomulka commented on distribution/src/config/log4j2.properties:
```diff
@@ -7,14 +7,37 @@ logger.action.level = debug
 appender.console.name = console
 appender.console.layout.type = PatternLayout
-appender.console.layout.pattern = [%d{ISO8601}][%-5p][%-25c{1.}] [%node_name]%marker %m%n
+appender.console.layout.pattern ={\
```
as per comment on the [issue](https://github.com/elastic/elasticsearch/issues/32850#issuecomment-448661944) should we allow users to change that pattern? If so, how do we validate that the JSON format is not broken

pgomulka replied:
That was solved by creating ESJsonLayout which users won't be able to change

pgomulka committed:
adding new option with clustestatelistener, pattern converter and threadlocal

removing worse options

pgomulka commented:
> There are two other options that I will try investigate:
> one would be to Modify and Reload log4j configuration with a new static pattern containing the node_id and clusterId once the ClusteStateChange is received.
That is actually hard to do. We can modify the configuration on the fly, but we would have to rewrite the whole config to Java. This is not easy to modify just a pattern. We also don't know what will be in user's log4j2.properties file. 
All these modified changes will also be gone if someone modifies the log4j2.properties in runtime (it will trigger refresh)

> another would be to write a custom formatter that will be checking MDC for presence of the node_id&cluster_id. If it was not yet set, it would try to get it from a shared AtomicReference and set in the MDC for future uses.
That option is actually best promising. No need to use MDC, we can just use the regular PatternConverter and a ThreadLocal. Should have the best performance from all suggested options, I haven't measured that yet though.

I have added a new commit with that best option only, removing other options. You can see these in previous commit.

pgomulka committed:
exception handling

pgomulka committed:
json message and exception works

pgomulka committed:
common wrapper layout class

ycombinator commented on server/src/main/java/org/elasticsearch/common/logging/ESJsonLayout.java:
```diff
@@ -0,0 +1,86 @@
+
+    private static final String PATTERN = "{" +
+        "\"type\": \"console\", " +
+        "\"timestamp\": \"%d{ISO8601}\", " +
```
Currently, all Elasticsearch logs except the GC log emit timestamps without the timezone. This is valid ISO8601, of course, but it would be nice if the time zone could be included as well. 

I see here that the `ESJsonLayout` will also emit ISO8601 timestamps. Will these include timezone information as well?

pgomulka replied:
@ycombinator I think this is worth adding - thanks for suggestion. Will include that in this issue

pgomulka committed:
removed debugging code

pgomulka committed:
passing dignity test

pgomulka committed:
fixing build

pgomulka committed:
Merge branch 'master' into feature/logging-structured

pgomulka committed:
fix failing test

pgomulka commented:
@elasticmachine retest this please

pgomulka committed:
fix failing test

pgomulka committed:
fix import

pgomulka committed:
extending logs test

pgomulka commented:
run the gradle build tests 1

pgomulka commented:
run the gradle build tests 1

pgomulka committed:
fix parsing and exception formatting

pgomulka committed:
fix failing test

pgomulka committed:
fix checkstyle

pgomulka committed:
Merge branch 'master' into feature/logging-structured

pgomulka committed:
small cleanup

pgomulka committed:
json logs cleanup

pgomulka committed:
Merge branch 'master' into feature/logging-structured

pgomulka committed:
test cleanup

ignore failing compilation error
refactor test not to use dropwhile
fix failing test
adding debuging info to failing test

pgomulka committed:
sometimes HttpServerTransport is logging first, and then the server dies. Not logging the exception from ElasticsearchUncaughtExceptionHandler

pgomulka committed:
additional json tests

pgomulka committed:
docker log4j config cleanup

pgomulka commented:
run the gradle build tests 2

pgomulka committed:
incorrect docker appender ref

pgomulka committed:
the right order of reading values from clusterListener

pgomulka committed:
add missing marker in a pattern

pgomulka committed:
empty lines cleanup

nik9000 commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogs.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class JsonLogs implements Iterable<JsonLogLine>, Closeable {
```
Could use javadoc I think.

nik9000 replied:
I think `Iterable`s have to support calling `iterator` multiple times to return independent `Iterator`. I don't see it in the Javadocs for `Iterable` but I'd be confused by an `Iterable` that works like this one. It might be more kind to have it implement `Stream` because that has `close` and "run once" semantics.

pgomulka replied:
agree -that needs a javadoc & and a cleanup.
The main intention is to use that class for readable asertions for json logger tests
That is why it has to be Iterable. You could think of this class as stream or Scanner, but neither of those has support in hamcrest at the version we use.

It can't only be stream, as we have some tests at the moment where the logs is traversed, some decision is made and then continued to be traversed from that point. We can't do that in stream (run once)

Refactored this to use a Reader supplier. Every time a new stream or iterator is created, a new Reader is created from the supplier.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsIntegTestCase.java:
```diff
@@ -0,0 +1,122 @@
+     * just check the entire log file because it could be quite long and
+     * exceptions don't include the node name.
+     */
+    private static final int LINES_TO_CHECK = 10;
```
Now that this class isn't explicitly about node names I'd change the name of this constant.

pgomulka replied:
the name is good, but the comment needs updating

nik9000 commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogLine.java:
```diff
@@ -0,0 +1,153 @@
+
+import java.util.List;
+
+public class JsonLogLine {
```
This class could also use javadoc I think. The methods probably don't need it though.

nik9000 commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -0,0 +1,94 @@
+
+@Plugin(category = PatternConverter.CATEGORY, name = "NodeAndClusterIdConverter")
+@ConverterKeys({"node_and_cluster_id"})
+public final class NodeAndClusterIdConverter extends LogEventPatternConverter implements ClusterStateListener {
```
I think rather than making the pattern converted a cluster state listener, I might make a cluster state listener that writes to a static method in the pattern converted. That'd keep Elasticsearch concepts out of the log4j extension code and wouldn't force leaking "funny" log4j2 constructs like `newInstance` into Elasticsearch.

pgomulka replied:
Ineed that would be a good design choice to separate the cluster state listener and the pattern converter. 
As discussed the static atomicReference field is safe here, as we only expect one instance of PatternConverter per jvm.

nik9000 commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -0,0 +1,94 @@
+
+    @Override
+    public void format(LogEvent event, StringBuilder toAppendTo) {
+        if (nodeAndClusterIds.get() != null) {
```
I think it is worth a comment about how this `ThreadLocal` is a cache for the `AtomicReference`. I do wonder, given the benchmarks that you have, if it is worth having the cache at all. It'd be nice to have some perf data for ARM too, just because I suspect it'd be quite different than x86.

pgomulka replied:
The comment is definetly worth putting - will do
As I mentioned in our discussion I am wondering if we need that optimization at all. The microbenchmarks I did for that class show that it won't benefit much (it is actually a bit slower)
Mikrobenchmark code&result:
https://github.com/pgomulka/elasticsearch/commit/dc13f47e45ca64910f50eff2e200194de30e4a5b

pgomulka replied:
The thread local field has been removed

nik9000 reviewed (commented):
I'm pretty sure I'm not the right person to *approve* this, but I did leave some review comments that I hope will be helpful. Thanks for looking into this! There is a lot going on here!

pgomulka committed:
Merge branch 'master' into feature/logging-structured

danielmitterdorfer reviewed (changes requested):
I did a first pass and left some comments.

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/ESJsonLayout.java:
```diff
@@ -0,0 +1,106 @@
+        "\"timestamp\": \"%d{yyyy-MM-dd'T'HH:mm:ss,SSSZ}\", " +
+        "\"level\": \"%p\", " +
+        "\"component\": \"%c{1.}\", " +
+        "\"cluster.name\": \"${sys:es.logs.cluster_name}\", " +
```
We should avoid dots in field names; this property should be called `cluster_name`.

pgomulka replied:
as per https://github.com/elastic/ecs#why-does-ecs-use-a-dot-notation-instead-of-an-underline-notation
and since this is pared by Beats shouldn't this be a dot notation?

danielmitterdorfer replied:
Makes sense. Let's keep the dot notation.

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/ESJsonLayout.java:
```diff
@@ -0,0 +1,106 @@
+        "\"level\": \"%p\", " +
+        "\"component\": \"%c{1.}\", " +
+        "\"cluster.name\": \"${sys:es.logs.cluster_name}\", " +
+        "\"node.name\": \"%node_name\", " +
```
We should avoid dots in field names; this property should be called `node_name`.

pgomulka committed:
addressing Nik's comments

pgomulka commented on qa/evil-tests/src/test/java/org/elasticsearch/common/logging/JsonLoggerTests.java:
```diff
@@ -0,0 +1,238 @@
+        Logger prefixLogger = new PrefixLogger(LogManager.getLogger("prefixLogger"), "PREFIX");
+        prefixLogger.info("This is an info message with a prefix");
+
+        final String path = System.getProperty("es.logs.base_path") +
```
good spot - fixed

pgomulka commented on qa/evil-tests/src/test/java/org/elasticsearch/common/logging/JsonLoggerTests.java:
```diff
@@ -0,0 +1,238 @@
+
+    @SuppressWarnings("unchecked")
+    public void testJsonLayout() throws IOException, UserException {
+        setupLogging("json_layout");
```
I have to say I followed blindly the convention from other tests in this package (copy pasted too much..) This would be worthy to fix there too, but don't want this PR to grow too much

danielmitterdorfer replied:
I think as this is a new test and the change is pretty straightforward we should do it here?

pgomulka commented on qa/evil-tests/src/test/java/org/elasticsearch/common/logging/JsonLoggerTests.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    private void setupLogging(final String config, final Settings settings) throws IOException, UserException {
+        assert !Environment.PATH_HOME_SETTING.exists(settings);
```
This again is a 'pattern' from other tests, but agree that junit assertion would be better with a message here

danielmitterdorfer replied:
I was rather thinking of something along the lines of:

```
assert Environment.PATH_HOME_SETTING.exists(settings) == false : "path.home must be absent in [" + settings + "]";

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/ESJsonLayout.java:
```diff
@@ -0,0 +1,106 @@
+     * exceptionAsJson - in json as stacktrace. Only present when throwable is passed as a parameter to a Logger. Taken from
+     * JsonThrowablePatternConverter
+     */
+    private static final String PATTERN = "{" +
```
There is no need for us to change anything in code if a user wants to update his pattern. This class does very little, only wraps the pattern that would normally be in a log4j2.properties file

So if a user wants to use a previous style of logs, or his own patterns, he can just change that ina log4j2.properties file

danielmitterdorfer replied:
Ok, then let's keep this as is.

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/ESJsonLayout.java:
```diff
@@ -0,0 +1,106 @@
+     * JsonThrowablePatternConverter
+     */
+    private static final String PATTERN = "{" +
+        "\"type\": \"${TYPE}\", " +
```
That could be easily added with a nonEmpty conversion pattern, but I think it is better now when the pattern is the same in docker and other binaries.
Would this also make it easier for beats to parse logs the same way in both deployments?

danielmitterdorfer replied:
The problem that I see is that it makes the log files even noisier. But you have a point regarding the consistency in parsing the files.

pgomulka replied:
having this less noisy would be very beneficial. How about ordering? previously we did not have node.id , cluster.uuid and others. Maybe we could have the order:
type, timestamp, level, component, message, node.name, cluster.name, node.id, cluster.uiid
?

danielmitterdorfer replied:
As discussed offline, let's keep this consistent.

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/JsonThrowablePatternConverter.java:
```diff
@@ -0,0 +1,105 @@
+     * Private constructor. Parameters only used to configure wrapped throwablePatternConverter
+     */
+    private JsonThrowablePatternConverter(final Configuration config, final String[] options) {
+        super("CustomExtendedThrowable", "throwable", options, config);
```
a mistake indeed - refactored the class name and forgot about this one.. fixed

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -0,0 +1,94 @@
+        try {
+            return INSTANCE.getOrCompute();
+        } catch (Exception e) {
+            return null;
```
I was paranoid about this class being constructed only once - I have confirmed that and there is no need for this getOrCompute at all

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -0,0 +1,94 @@
+        boolean wasSet = nodeAndClusterIdsReference.compareAndSet(null, formatIds(clusterUUID, nodeId));
+
+        if (wasSet) {
+            logger.info("received first cluster state update. Setting nodeId={} and clusterUuid={}", nodeId, clusterUUID);
```
fixed

pgomulka commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -437,6 +439,9 @@ protected Node(
                                                  namedWriteableRegistry).stream())
                 .collect(Collectors.toList());
 
+            NodeAndClusterIdConverter nodeAndClusterIdConverter = NodeAndClusterIdConverter.newInstance(new String[]{});
```
removed at all

pgomulka commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsIntegTestCase.java:
```diff
@@ -0,0 +1,122 @@
+
+/**
+ * Tests that extend this class verify that all json layout fields appear in the first few log lines after startup
+ * Fields available straight away : type, timestamp, level, component, message, nodeName, clusterName
```
reworded

pgomulka commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsIntegTestCase.java:
```diff
@@ -0,0 +1,122 @@
+ * Fields available straight away : type, timestamp, level, component, message, nodeName, clusterName
+ * nodeId and clusterId are available later once the clusterState was received.
+ * NodeName, ClusterName, NodeId, ClusterId should be consistent across all log lines
+ * Note that this won't pass for clusters that don't have
```
These comments are about node.name as this class was previously testing node.name only
So when trace or debug is enabled and a elasticsearch.yml contains a node.name property, some log lines will then appear with that node.name BEFORE we set the node.name from LogConfigurator
I reworded that in a comment so hopefully it is more clear

pgomulka commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsIntegTestCase.java:
```diff
@@ -0,0 +1,122 @@
+ * Tests that extend this class verify that all json layout fields appear in the first few log lines after startup
+ * Fields available straight away : type, timestamp, level, component, message, nodeName, clusterName
+ * nodeId and clusterId are available later once the clusterState was received.
+ * NodeName, ClusterName, NodeId, ClusterId should be consistent across all log lines
```
Once they apear, they shoud not change in logs

pgomulka committed:
follow up after Daniel's comments

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -0,0 +1,94 @@
+
+    private final Logger logger = LogManager.getLogger(getClass());
+    private final AtomicReference<String> nodeAndClusterIdsReference = new AtomicReference<>();
+    private final CloseableThreadLocal<String> nodeAndClusterIds = new CloseableThreadLocal();
```
removed the thread local

pgomulka commented on distribution/src/config/log4j2.properties:
```diff
@@ -6,14 +6,15 @@ logger.action.level = debug
 
 appender.rolling.type = RollingFile
 appender.rolling.name = rolling
 appender.rolling.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}.log
```
almost forgot about this one - thanks for spotting
would this be better if we use the ${cluster_name}_server pattern
we already have this pattern for deprecated and other types of logs

danielmitterdorfer replied:
I guess a prefix would be easier to match than a suffix but as the Beats team needs to handle this, can you please check with them?

pgomulka committed:
failing test

pgomulka committed:
unused imports

pgomulka committed:
failing tests

pgomulka committed:
rename test log name

pgomulka committed:
method rename

pgomulka committed:
Merge branch 'master' into feature/logging-structured

pgomulka committed:
rename name to server

pgomulka committed:
rename revert and level corrected

pgomulka commented:
run the gradle build tests 1

pgomulka committed:
wrong assertion

pgomulka committed:
rename log name files in package tests

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/JsonThrowablePatternConverter.java:
```diff
@@ -0,0 +1,105 @@
+@Plugin(name = "JsonThrowablePatternConverter", category = PatternConverter.CATEGORY)
+@ConverterKeys({"exceptionAsJson"})
+public final class JsonThrowablePatternConverter extends ThrowablePatternConverter {
+
```
Personally I find the new line odd in this place but I checked the code base and we have no preferred style. So let's keep it as is.

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/ESJsonLayout.java:
```diff
@@ -33,21 +33,29 @@
-     * node_and_cluster_id - in json as node.id and cluster.uuid - taken from NodeAndClusterIdConverter and present
+     * Fields used in a pattern to format a json log line:
+     * <ul>
+     * <li>  type - the type of logs. These represent appenders and help docker distinguish log streams.
```
is this actually valid? I think the closing tags are missing, i.e. the structure needs to be `<li>Text</li>` instead of `<li>Text`. Also nit: Remove leading space for each item.

pgomulka replied:
It does work indeed, but I guess better to stick with the html format. Also the space removed.

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -19,76 +19,60 @@
 
-
+/**
+ * Pattern converter to format the node_and_cluster_id variable into a json fields node.id and cluster.uuid
```
Nits:

* `node.id`-> `<code>node.id</code>` (similarly for `cluster.uuid`)
* Missing punctuation mark

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -19,76 +19,60 @@
-
+/**
+ * Pattern converter to format the node_and_cluster_id variable into a json fields node.id and cluster.uuid
+ * Keeping those two fields together assures that the will be atomically set and become visible in logs at the same time
```
Nits:

* "that the" -> "that they"
* Missing punctuation mark

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -19,76 +19,60 @@
     /**
      * Called by log4j2 to initialize this converter.
      */
     public static NodeAndClusterIdConverter newInstance(final String[] options) {
```
Can you please add `@SuppressWarnings("unused")` to the parameter?

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdStateListener.java:
```diff
@@ -0,0 +1,47 @@
+ *  it sets the clusterUUID and nodeID in log4j pattern converter {@link NodeAndClusterIdConverter}
+ */
+public class NodeAndClusterIdStateListener implements ClusterStateListener {
+    private final Logger logger = LogManager.getLogger(getClass());
```
Any reason this is not declared `private static final`?

pgomulka replied:
just the matter of preference of using getClass() vs NodeAndClusterIdStateListener.class
there would be only one instance of this class (created in Node.java) so there should be no overhead in performance.

danielmitterdorfer replied:
My comment was more related to consistency than about performance. The preferred style in our code base is to use class constants so I think we should stick to that convention here as well.

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdStateListener.java:
```diff
@@ -0,0 +1,47 @@
+        boolean wasSet = NodeAndClusterIdConverter.setOnce(clusterUUID, nodeId);
+
+        if (wasSet) {
+            logger.info("Eeceived first cluster state update. Setting nodeId=[{}] and clusterUuid=[{}]", nodeId, clusterUUID);
```
Nit: `Eeceived` -> `Received`. I also wonder whether this is worth an info level message? Maybe just log this on debug level?

pgomulka replied:
You are right, it was useful for me when developing and debuging this PR, but indeed it would not be necessary to see every time by regular users. Will change to debug

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/JsonThrowablePatternConverter.java:
```diff
@@ -0,0 +1,106 @@
+@Plugin(name = "JsonThrowablePatternConverter", category = PatternConverter.CATEGORY)
+@ConverterKeys({"exceptionAsJson"})
+public final class JsonThrowablePatternConverter extends ThrowablePatternConverter {
+    private static final JsonStringEncoder JSON_STRING_ENCODER = JsonStringEncoder.getInstance();
```
I checked the implementation of `JsonStringEncoder`. It never mentions it explicitly but it does not appear to be thread-safe (given that `#getInstance()` maintains a thread-local). So I think it is not safe to store it as a field.

pgomulka replied:
great point, I didn't expect that. will refactor into local use

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -0,0 +1,78 @@
+@Plugin(category = PatternConverter.CATEGORY, name = "NodeAndClusterIdConverter")
+@ConverterKeys({"node_and_cluster_id"})
+public final class NodeAndClusterIdConverter extends LogEventPatternConverter {
+    private static final AtomicReference<String> nodeAndClusterIdsReference = new AtomicReference<>();
```
The plural in the name `nodeAndClusterIdsReference` implies that there are multiple node and cluster ids when there is only one. Also I think the suffix `reference` does not add meaning? So maybe just `nodeAndClusterId`? Also, do we want to declare this as plain final field instead of a class constant given that `NodeAndClusterIdConverter` is effectively a singleton (as per Log4j's contract)?

pgomulka replied:
agree the name change here would be beneficial - done

re - plain final field. Relying on the  fac that Log4j only instantiates this once makes it easier to pass the value from `NodeAndClusterIdStateListener.java` to this field with the use of static setOnce method.  
If this was a regular final field the cluster state listener would have to access that only instance used by log4j - therefore this class would have to implement a `newInstance` in a singleton style. This is how it was done previously and I think it was redundant.

danielmitterdorfer commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public NodeAndClusterIdConverter() {
+        super("NodeName", "node_and_cluster_id");
```
Should the first constructor parameter be `NodeAndClusterId` instead of `NodeName` as this class is handling both fields?

danielmitterdorfer commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsIntegTestCase.java:
```diff
@@ -0,0 +1,124 @@
+ */
+public abstract class JsonLogsIntegTestCase extends ESRestTestCase {
+    /**
+     * Number of lines in the log file to check for the node_name, node.id or cluster.uuid. We don't
```
I think `node_name` should not be mentioned in this comment? Also, I'd prefer to wrap names like `node.id` and `cluster.uuid` in `<code>` tags.

pgomulka replied:
this test is still is asserting about the cluster_name, so I think it still would be good to keep it

danielmitterdorfer replied:
You're right about the node name, I missed that before.

danielmitterdorfer commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsStream.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    public static Stream<JsonLogLine> from(Path path) throws IOException {
+        return new JsonLogsStream(Files.newBufferedReader(path)).stream();
```
```suggestion
        return from(Files.newBufferedReader(path));
```

pgomulka replied:
good suggestion - will add

danielmitterdorfer commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsStream.java:
```diff
@@ -0,0 +1,98 @@
+        return new JsonLogsStream(Files.newBufferedReader(path)).stream();
+    }
+
+    public Stream<JsonLogLine> stream() {
```
Can we declare this as `private` because it should only be called by the factory methods above?

danielmitterdorfer commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsStream.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    private class JsonIterator implements Iterator<JsonLogLine> {
```
can we declare this as `static` inner class and just pass `parser` to the iterator instead of implicitly capturing the whole state of the outer class?

danielmitterdorfer reviewed (changes requested):
I did another pass now.

pgomulka commented on qa/evil-tests/src/test/java/org/elasticsearch/common/logging/JsonLoggerTests.java:
```diff
@@ -0,0 +1,238 @@
+
+    @BeforeClass
+    public static void initNodeName() {
+        LogConfigurator.setNodeName("sample-name");
```
That is a very good point. There is another test in this module that calls that, but somehow it does not throw an exception here..
Need to investiate that

pgomulka replied:
that inded fails if run with `-Dtests.jvms=` . I created a separate module in QA with this test moved there and additional test that verifies that previous configuration style will still work

pgomulka committed:
addressing Daniels' second round of comments

danielmitterdorfer reviewed (changes requested):
I had yet another pass and I think things are in rather good shape. I still think we need to work on docs (as previously mentioned).

pgomulka committed:
javadocs

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -23,40 +23,42 @@ will resolve to `/var/log/elasticsearch/production.log`.
 <1> Configure the `RollingFile` appender
 <2> Log to `/var/log/elasticsearch/production.log`
-<3> Roll logs to `/var/log/elasticsearch/production-yyyy-MM-dd-i.log`; logs
+<3> Use Json layout.
```
JSON is an acronym. Hence it is "JSON" not "Json".

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -23,40 +23,42 @@ will resolve to `/var/log/elasticsearch/production.log`.
-appender.rolling.filePattern = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}-%d{yyyy-MM-dd}-%i.log.gz <3>
+appender.rolling.layout.type = ESJsonLayout <3>
+appender.rolling.layout.type_name = server <4>
+appender.rolling.filePattern = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}-%d{yyyy-MM-dd}-%i.log.gz <5>
```
According to our new default config this should include the `_server` suffix. I think you should also align the example file name in the paragraph above with the file name here, i.e. in the end the name should be `/var/log/elasticsearch/production_server.log`.

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -23,40 +23,42 @@ will resolve to `/var/log/elasticsearch/production.log`.
-<3> Roll logs to `/var/log/elasticsearch/production-yyyy-MM-dd-i.log`; logs
+<3> Use Json layout.
+<4> `type_name` is a flag populating `type` field in a `ESJsonLayout`. It is used to distinguish streams of logs when parsing.
+<5> Roll logs to `/var/log/elasticsearch/production-yyyy-MM-dd-i.log`; logs
```
According to our new default config this should include the _server suffix.

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -23,40 +23,42 @@ will resolve to `/var/log/elasticsearch/production.log`.
 <2> Log to `/var/log/elasticsearch/production.log`
-<3> Roll logs to `/var/log/elasticsearch/production-yyyy-MM-dd-i.log`; logs
+<3> Use Json layout.
+<4> `type_name` is a flag populating `type` field in a `ESJsonLayout`. It is used to distinguish streams of logs when parsing.
```
Nit: "populating *the* `type` field". 

How about "It is used to distinguish streams of logs when parsing." -> "It can be used to distinguish different types of logs more easily when parsing them."?

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+
+[float]
+[[json-logging]]
+=== Json log format
```
Json -> JSON

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+[[json-logging]]
+=== Json log format
+
+To make parsing elasticsearch logs easier, logs are now printed in a json format.
```
elasticsearch -> Elasticsearch
json -> JSON

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+=== Json log format
+
+To make parsing elasticsearch logs easier, logs are now printed in a json format.
+This is configured by a log4j layout property `appender.rolling.layout.type = ESJsonLayout`.
```
log4j -> Log4J

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+
+To make parsing elasticsearch logs easier, logs are now printed in a json format.
+This is configured by a log4j layout property `appender.rolling.layout.type = ESJsonLayout`.
+This layout requires a type_name attribute to be set which is used to distinguish
```
nit: enclose type_name in backticks.

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+appender.rolling.layout.type_name = server
+--------------------------------------------------
+
+Each line contains a single json with set of attributes that are configured in `ESJsonLayout`.
```
```suggestion
Each line contains a single JSON document with the properties configured in `ESJsonLayout`.
```

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+--------------------------------------------------
+
+Each line contains a single json with set of attributes that are configured in `ESJsonLayout`.
+See this class javadoc for more details.
```
I am not sure we should point users of our reference docs to the Javadoc. In the end it would mean that they'd either need to checkout the sources or go to Github, find the right class and read the Javadocs which seems inconvenient. I wonder whether we should go the extra mile and document the properties here as well although the information is duplicated then. Or we don't document it in Javadoc and only in the reference docs. Either way, I think we can improve this

pgomulka replied:
or maybe I could point them directly?
We do in sam documentation pages already. For isntance: https://www.elastic.co/guide/en/elasticsearch/client/java-rest/6.0/java-rest-high-query-builders.html

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+
+Each line contains a single json with set of attributes that are configured in `ESJsonLayout`.
+See this class javadoc for more details.
+When the log event contains a `Throwable`, json will be formatted multiline where first line contains
```
I think this too technical. How about: "Exception stack traces will be formatted as multiple lines ..."

pgomulka replied:
`However if JSON documents contains exception, stack trace will be formatted as a multiline.
`

pgomulka replied:
will reword

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+Each line contains a single json with set of attributes that are configured in `ESJsonLayout`.
+See this class javadoc for more details.
+When the log event contains a `Throwable`, json will be formatted multiline where first line contains
+all the regular attributes and subsequent lines will contain stacktrace formatted as a json array.
```
I think the term to use is "properties", not "attributes" but I might be wrong. Also: json -> JSON

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+all the regular attributes and subsequent lines will contain stacktrace formatted as a json array.
+
+
+NOTE: You can still use your own custom layout. To do that just replace line
```
Scratch "just"? It does not convey any meaning here IMHO? Also: "replace line" -> replace *the* line"

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+
+
+NOTE: You can still use your own custom layout. To do that just replace line
+`appender.rolling.layout.type` with the layout of your preference. See sample below:
```
"with the layout of your preference" -> "with a different layout" or "with your preferred layout"?

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+--------------------------------------------------
+appender.rolling.type = RollingFile
+appender.rolling.name = rolling
+appender.rolling.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}.log
```
Add `_server` suffix for consistency?

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +196,36 @@ files (four rolled logs, and the active log).
+appender.rolling.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}.log
+appender.rolling.layout.type = PatternLayout
+appender.rolling.layout.pattern = [%d{ISO8601}][%-5p][%-25c{1.}] [%node_name]%marker %.-10000m%n
+appender.rolling.filePattern = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}-%d{yyyy-MM-dd}-%i.log.gz
```
Add `_server` suffix for consistency?

pgomulka replied:
as discussed not adding to archive names

danielmitterdorfer commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsIntegTestCase.java:
```diff
@@ -35,13 +35,16 @@
  *
  * Note that this won't pass for nodes in clusters that don't have the node name defined in elasticsearch.yml <strong>and</strong> start
- * with DEBUG or TRACE level logging. Those nodes log a few lines before the node.name is set by LogConfigurator.setNodeName
+ * with DEBUG or TRACE level logging. Those nodes log a few lines before the node.name is set by <code>LogConfigurator.setNodeName</code>
```
Nit: Missing punctuation mark.

danielmitterdorfer commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsIntegTestCase.java:
```diff
@@ -35,13 +35,16 @@
- * whereas: nodeId and clusterId are available later once the clusterState was received.
+ * Fields available from the first log line: <code>type</code>, <code>timestamp</code>, <code>level</code>, <code>component</code>,
+ * <code>message</code>, <code>node.name</code>, <code>cluster.name</code>.
+ * Whereas <code>node.id</code> and <code>cluster.uuid</code> are available later once the clusterState was received.
```
How about ".... once the *first* cluster state update has been received"?

danielmitterdorfer commented on test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogsIntegTestCase.java:
```diff
@@ -35,13 +35,16 @@
  * Tests that extend this class verify that all json layout fields appear in the first few log lines after startup
- * Fields available from the first log line: type, timestamp, level, component, message, nodeName, clusterName
- * whereas: nodeId and clusterId are available later once the clusterState was received.
+ * Fields available from the first log line: <code>type</code>, <code>timestamp</code>, <code>level</code>, <code>component</code>,
```
How about "from the first log line" -> "upon process startup"?

danielmitterdorfer reviewed (changes requested):
Thanks for adding docs. I left a few suggestions.

pgomulka committed:
additional test verifing old config

move jsonLoggerTest to separate module to avoid clash with EvilLoggerTests

pgomulka committed:
unused import

pgomulka committed:
empty unused test

pgomulka committed:
small fixes after review

danielmitterdorfer commented on qa/logging-config/build.gradle:
```diff
@@ -25,7 +25,10 @@ apply plugin: 'elasticsearch.standalone-test'
     autoSetHostsProvider = false
-    // overriding the distribution log4j configuration file with a custom configuration
+    /**
+     * The intention is to provide a custom log4j configuration where layout is an old style pattern and confirm that Elasticsearch 
```
I think it is fine to just start the sentence with "Provide a custom ..."

danielmitterdorfer commented on qa/logging-config/build.gradle:
```diff
@@ -25,7 +25,10 @@ apply plugin: 'elasticsearch.standalone-test'
-    // overriding the distribution log4j configuration file with a custom configuration
+    /**
+     * The intention is to provide a custom log4j configuration where layout is an old style pattern and confirm that Elasticsearch 
+     * can successfully startup. see CustomLoggingIT.
```
It is not `CustomLoggingIT` but `CustomLoggingConfigIT`. Anyway, I don't think we need to mention it?

danielmitterdorfer reviewed (commented):
I left a couple more minor comments

pgomulka committed:
comment cleanup after review

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +197,38 @@ files (four rolled logs, and the active log).
+
+Each line contains a single JSON document with the properties configured in `ESJsonLayout`.
+See this class {es-json-layout-java-doc}[javadoc] for more details.
+However if a JSON document contains exception, it will be formatted as a multiline.
```
Nits: 
* "contains **an** exception"
* "formatted as a multiline" -> "printed over multiple lines"

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +197,38 @@ files (four rolled logs, and the active log).
+Each line contains a single JSON document with the properties configured in `ESJsonLayout`.
+See this class {es-json-layout-java-doc}[javadoc] for more details.
+However if a JSON document contains exception, it will be formatted as a multiline.
+First line will contain regular properties and subsequent lines will contain
```
Nit: "**The** first line"

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +197,38 @@ files (four rolled logs, and the active log).
+See this class {es-json-layout-java-doc}[javadoc] for more details.
+However if a JSON document contains exception, it will be formatted as a multiline.
+First line will contain regular properties and subsequent lines will contain
+stacktrace formatted as a JSON array.
```
Nit: "**the** stacktrace"

danielmitterdorfer commented on server/src/test/java/org/elasticsearch/common/logging/JsonThrowablePatternConverterTests.java:
```diff
@@ -0,0 +1,92 @@
+import java.io.StringReader;
+
+import static org.hamcrest.Matchers.equalTo;
+/*
```
The license header should be at the top of the file instead of somewhere in the middle.

danielmitterdorfer reviewed (commented):
I did one further pass and left a few nits.

pgomulka committed:
documentation and licence fix

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -216,8 +216,8 @@ appender.rolling.layout.type_name = server
 See this class {es-json-layout-java-doc}[javadoc] for more details.
-However if a JSON document contains exception, it will be formatted as a multiline.
-First line will contain regular properties and subsequent lines will contain
+However if a JSON document contains an exception, it will be printted over multiple lines.
```
typo: "printted" -> "printed"

danielmitterdorfer reviewed (commented):
Almost there. I found one typo.

pgomulka committed:
typo printted -> printed

danielmitterdorfer reviewed (approved):
Thanks for iterating! LGTM

pgomulka commented:
run the gradle build test 1

pgomulka commented:
run the gradle build test 2

pgomulka commented:
run the gradle build tests 2

pgomulka commented:
run the gradle build tests 1

rjernst reviewed (commented):
I left a few comments. Also, could you please post a comment here with examples of log lines in the new log, both a normal line, and one with an exception? I'm concerned with readability if the entire exception is in one line (maybe I misread the way it would work?), and also with alignment when skimming through logs as right now several elements are padded for alignment but it doesn't look like anything is in the new format?

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java:
```diff
@@ -0,0 +1,78 @@
+@Plugin(category = PatternConverter.CATEGORY, name = "NodeAndClusterIdConverter")
+@ConverterKeys({"node_and_cluster_id"})
+public final class NodeAndClusterIdConverter extends LogEventPatternConverter {
+    private static final AtomicReference<String> nodeAndClusterId = new AtomicReference<>();
```
SetOnce makes total sense in `NodeNamePatternConverter` as we call it just once during log configuration. However it is throwing exception if there is an attempt to set it again. 

As this field is set by a  `ClusteStateListener`, there will be more attempts to call setOnce method later.
We can consider deregister the `NodeAndClusterIdStateListener` once it received the first update, but there still is a possibility of two cluster state updates one after another. So we will still need to guard against multiple calls to SetOnce. What do you think?
@danielmitterdorfer any views on this?

danielmitterdorfer replied:
`SetOnce` does make sense to me for this use case but I am bit worried about the complexity that we'd introduce to correctly unsubscribe from concurrent cluster state updates.

pgomulka commented on qa/logging-config/src/test/resources/plugin-security.policy:
```diff
@@ -0,0 +1,4 @@
+grant {
+  // Needed to read the log file
+  permission java.io.FilePermission "${tests.logfile}", "read";
```
It is the integrationTestRunner that needs that. Otherwise we will get:
```
REPRODUCE WITH: ./gradlew :qa:logging-config:integTestRunner -Dtests.seed=6D8D42254DAD541C -Dtests.class=org.elasticsearch.qa.custom_logging.CustomLoggingConfigIT -Dtests.method="testSuccessfulStartupWithCustomConfig" -Dtests.security.manager=true -Dtests.locale=en-IN -Dtests.timezone=Asia/Tel_Aviv -Dtests.distribution=zip -Dcompiler.java=11 -Druntime.java=8
ERROR   1.12s | CustomLoggingConfigIT.testSuccessfulStartupWithCustomConfig <<< FAILURES!
   > Throwable #1: java.security.AccessControlException: access denied ("java.io.FilePermission" "/Users/przemek/workspace/pgomulka/elasticsearch/qa/logging-config/build/cluster/integTestCluster node0/elasticsearch-7.0.0-SNAPSHOT/logs/qa_logging-config_integTestCluster_server.log" "read")
   >    at __randomizedtesting.SeedInfo.seed([6D8D42254DAD541C:F2069750FBE5E1B8]:0)
   >    at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)

```

pgomulka committed:
setOnce argument ordering

test pattern simplification

pgomulka committed:
javadoc typo

dliappis commented on distribution/docker/src/docker/config/log4j2.properties:
```diff
@@ -1,9 +1,44 @@
+appender.index_indexing_slowlog_rolling.layout.type = ESJsonLayout
+appender.index_indexing_slowlog_rolling.layout.type_name = index_indexing_slowlog
+
+
```
nit: not need for an extra newline here

dliappis reviewed (approved):
LGTM; the Docker config looks great and seems to be working very well. I left a minor comment.

pgomulka committed:
refactor cluster state listeners to use setOnce

pgomulka committed:
removed empty line

pgomulka committed:
methods rename and cleanup

pgomulka committed:
javadoc typo

pgomulka committed:
Merge pull request #7 from pgomulka/fix/observer-logging

Refactor to use setOnce  and observable

pgomulka committed:
keep the old appenders and let the nodeIDlistener start earlier

pgomulka committed:
Merge branch 'master' into feature/logging-structured

danielmitterdorfer commented on docs/reference/setup/logging-config.asciidoc:
```diff
@@ -194,3 +197,38 @@ files (four rolled logs, and the active log).
+stacktrace formatted as a JSON array.
+
+
+NOTE: You can still use your own custom layout. To do that replace the line
```
I think the docs need to take into account that we now log as JSON *and* in plain text.

danielmitterdorfer commented on distribution/src/config/log4j2.properties:
```diff
@@ -29,58 +31,144 @@ appender.rolling.strategy.action.condition.type = IfFileName
 appender.deprecation_rolling.strategy.type = DefaultRolloverStrategy
 appender.deprecation_rolling.strategy.max = 4
+##############################################
+######## Rolling -  old style pattern ########
```
This should probably say `Deprecation - old style pattern`

danielmitterdorfer commented on distribution/src/config/log4j2.properties:
```diff
@@ -29,58 +31,144 @@ appender.rolling.strategy.action.condition.type = IfFileName
 appender.rolling.strategy.action.condition.nested_condition.type = IfAccumulatedFileSize
 appender.rolling.strategy.action.condition.nested_condition.exceeds = 2GB
+##############################################
+######## Server -  old style pattern ########
```
Very minor nit: Is it intentional that the comments for the old style are different than for JSON? Also, is it intentional that the first commen line has a different length than the second one?

pgomulka replied:
I will realign them, these should be same length 50.

danielmitterdorfer commented on distribution/src/config/log4j2.properties:
```diff
@@ -29,58 +31,144 @@ appender.rolling.strategy.action.condition.type = IfFileName
+appender.rolling_old.name = rolling_old
+appender.rolling_old.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}_server.old_log
+appender.rolling_old.layout.type = PatternLayout
+appender.rolling_old.layout.pattern = [%d{ISO8601}][%-5p][%-25c{1.}] [%node_name]%marker %m%n
```
I'd say it is ok that we keep the old format as is but now the new format contains different information (e.g. the cluster UUID is only present in the JSON format).

danielmitterdorfer reviewed (commented):
I think we need to adapt the release note docs but apart from that I only found very minor things.

danielmitterdorfer commented:
> I think we need to adapt the release note docs but apart from that I only found very minor things.

I stand corrected. The docs I've commented on are actually the logging docs. I think we need to add an entry to the release notes as this is a breaking change.

pgomulka committed:
improved documentation and more robust test

pgomulka committed:
Merge branch 'master' into feature/logging-structured

pgomulka committed:
split logging config in 2 for docs

pgomulka committed:
enable log print out for this test

pgomulka commented:
@elasticmachine run elasticsearch-ci/1

pgomulka committed:
rename logs to .json

pgomulka commented:
@elasticmachine run elasticsearch-ci/1

pgomulka commented:
@elasticmachine run elasticsearch-ci/1

pgomulka committed:
migration logging

pgomulka committed:
old log rename and documentation update

pgomulka committed:
making test more stable

deprecation docs

danielmitterdorfer commented on docs/reference/migration/migrate_7_0/logging.asciidoc:
```diff
@@ -3,8 +3,33 @@
+* ${cluster_name}_index_indexing_slowlog.log
+* ${cluster_name}_audit.json
+* gc.log
+* http.ports
```
We should not mention the `*.ports` files; they won't get created in production.

danielmitterdorfer commented on docs/reference/migration/migrate_7_0/logging.asciidoc:
```diff
@@ -3,8 +3,33 @@
+* http.ports
+* transport.ports
+
+Note: You can override additional files by removing the unwanted appenders in `log4j2.properties`.
```
Maybe: "You can configure which of these files are written by editing `log4j2.properties`"?

danielmitterdorfer commented on docs/reference/migration/migrate_7_0/logging.asciidoc:
```diff
@@ -3,8 +3,33 @@
+[float]
+==== Log files ending with `*.log` deprecated
+
+Log files with `*.log` using old pattern layout format are now considered deprecated
```
"Log files with the `.log` file extension using the old pattern layout format ..."

danielmitterdorfer commented on docs/reference/migration/migrate_7_0/logging.asciidoc:
```diff
@@ -3,8 +3,33 @@
+==== Log files ending with `*.log` deprecated
+
+Log files with `*.log` using old pattern layout format are now considered deprecated
+and newly added JSON logs ending `*.json` should be used.
```
"and the newly added JSON log file format with the `.json` file extension should be used instead."

danielmitterdorfer commented on docs/reference/migration/migrate_7_0/logging.asciidoc:
```diff
@@ -3,8 +3,33 @@
+
+Log files with `*.log` using old pattern layout format are now considered deprecated
+and newly added JSON logs ending `*.json` should be used.
+Note: GC logs in a file `gc.log` are not going to be changed.
```
"GC logs which are written to the file `gc.log` will not be changed"?

danielmitterdorfer commented on docs/reference/migration/migrate_7_0/logging.asciidoc:
```diff
@@ -3,8 +3,33 @@
+[float]
+==== Docker output in JSON format
+
+All docker console logs are now in JSON format. You can distinguish logs streams with a `type` field.
```
* docker -> Docker
* "... distinguish logs streams with **the** `type` field"

danielmitterdorfer commented on qa/logging-config/custom-log4j2.properties:
```diff
@@ -7,7 +7,7 @@ logger.action.level = debug
 appender.rolling.type = RollingFile
 appender.rolling.name = rolling
-appender.rolling.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}_server.json
+appender.rolling.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}_server.log
```
I think the `_server` suffix needs to be removed as well for the `.log` file?

pgomulka replied:
this is a test file. I intentionally keep this file slightly different

danielmitterdorfer reviewed (commented):
I've left a few comments on the new commits.

pgomulka committed:
doc changes after review

danielmitterdorfer commented on docs/reference/migration/migrate_7_0/logging.asciidoc:
```diff
@@ -17,19 +17,17 @@ Following files should be expected now in log directory:
-Note: GC logs in a file `gc.log` are not going to be changed.
+Log files with the `.log` file extension using the old pattern layout format
+are now considered deprecated and the newly added JSON log file format with
+the .json file extension should be used instead.
```
nit: formatting: .json should be enclosed in backticks.

danielmitterdorfer commented on docs/reference/migration/migrate_7_0/logging.asciidoc:
```diff
@@ -17,19 +17,17 @@ Following files should be expected now in log directory:
+Log files with the `.log` file extension using the old pattern layout format
+are now considered deprecated and the newly added JSON log file format with
+the .json file extension should be used instead.
+Note: GC logs which are written to the file gc.log will not be changed.
```
nit: formatting: gc.log should be enclosed in backticks?

danielmitterdorfer reviewed (commented):
I have two nits but apart from that it looks fine.

pgomulka committed:
fix doc

pgomulka committed:
Merge branch 'master' into feature/logging-structured

danielmitterdorfer reviewed (approved):
LGTM

pgomulka commented:
@elasticmachine run elasticsearch-ci/1

pgomulka committed:
Merge branch 'master' into feature/logging-structured

pgomulka commented:
@elasticmachine run elasticsearch-ci/1

romseygeek commented:
We're seeing quite a few test failures with a stack trace from this commit, looking like so:
```
1> [2019-01-31T04:23:38,262][WARN ][o.e.c.s.ClusterApplierService] [node_t0] failed to notify ClusterStateListener
  1> org.apache.lucene.util.SetOnce$AlreadySetException: The object cannot be set twice!
  1> 	at org.apache.lucene.util.SetOnce.set(SetOnce.java:69) ~[lucene-core-8.0.0-snapshot-83f9835.jar:8.0.0-snapshot-83f9835 83f9835a47a00a2ec58a4cf5fc0d492497cf7898 - jpountz - 2019-01-21 13:06:00]
  1> 	at org.elasticsearch.common.logging.NodeAndClusterIdConverter.setNodeIdAndClusterId(NodeAndClusterIdConverter.java:59) ~[main/:?]
  1> 	at org.elasticsearch.common.logging.NodeAndClusterIdStateListener.onNewClusterState(NodeAndClusterIdStateListener.java:69) ~[main/:?]
  1> 	at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onNewClusterState(ClusterStateObserver.java:308) ~[main/:?]
  1> 	at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.clusterChanged(ClusterStateObserver.java:193) ~[main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService.lambda$callClusterStateListeners$6(ClusterApplierService.java:481) ~[main/:?]
  1> 	at java.util.concurrent.ConcurrentHashMap$KeySpliterator.forEachRemaining(ConcurrentHashMap.java:3527) [?:1.8.0_202]
  1> 	at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:743) [?:1.8.0_202]
  1> 	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580) [?:1.8.0_202]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListeners(ClusterApplierService.java:478) [main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:467) [main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:414) [main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:165) [main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681) [main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [main/:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_202]
  1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_202]
  1> 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_202]
```
It's not obvious if this is the cause of the failures (which are in a number of different tests) or if it's just happening anyway and is adding extra noise to the logs, but either way it's worth investigating.

pgomulka commented:
@romseygeek this should not be causing your tests to fail, but definetely should be fixed
I have a PR for this:
#38110
